### DISCUSSION TITLE
Nullable with generic types behaving weird

### DISCUSSION THREAD

**Discussion by Cyl18** [dotnet/csharplang#3912] [General] [answered]
![image](https://user-images.githubusercontent.com/14993992/93616356-38083b00-fa07-11ea-983b-647ed3e97be5.png)
if I add a '?' to the TValue
![image](https://user-images.githubusercontent.com/14993992/93616440-52daaf80-fa07-11ea-8e7e-a78edc1fac7b.png)
![image](https://user-images.githubusercontent.com/14993992/93616478-5c641780-fa07-11ea-83e7-0ced7162ebbb.png)

so how can I improve this method?

**Comment by 333fred** [comment, marked as answer]
That latter code will be valid in C# 9.

**Reply by IS4Code** [reply]
You mean `TValue?` will be valid without constraints on the parameter? How is it going to be implemented?

**Reply by CyrusNajmabadi** [reply]
Unconstrained `TValue?` just means "calling this may result in getting `default(TValue)` back"

**Reply by 333fred** [reply]
> You mean `TValue?` will be valid without constraints on the parameter? How is it going to be implemented?

And, to be sure, it's already implemented and has been for a few months. You can go check it out right now.

**Comment by Joe4evr** [comment]
Out of curiosity, is there any reason you're not using `TryGetValue`? Because right now you're doing a double-lookup, which could hurt your perf quite a bit.

**Reply by Cyl18** [reply]
thank you! I didn't know that
